sw/device/lib/dif/dif_usbdev.h

```diff
@@ -0,0 +1,772 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_USBDEV_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_USBDEV_H_
+
+#include "sw/device/lib/base/mmio.h"
+#include <stddef.h>
+#include <stdint.h>
+
+/**
+ * Hardware constants.
+ */
+#define USBDEV_NUM_ENDPOINTS 12
+#define USBDEV_MAX_PACKET_SIZE 64
+// Internal constant that should not be used by clients. Defined here because
+// it is used in the definition of `dif_usbdev_buffer_pool` below.
+#define USBDEV_NUM_BUFFERS 32
+
+/**
+ * Free buffer pool.
+ *
+ * A USB device has a fixed number of buffers that are used for storing incoming
+ * and outgoing packets and the software is responsible for keeping track of
+ * free buffers. The pool is implemented as a stack for constant-time add and
+ * remove. `top` points to the last free buffer added to the pool. The pool is
+ * full when `top == USBDEV_NUM_BUFFERS - 1` and empty when `top == -1`.
+ */
+typedef struct dif_usbdev_buffer_pool {
+  uint8_t entries[USBDEV_NUM_BUFFERS];
+  int8_t top;
```

Comment:
Wouldn't it make more sense for this to be named `len`, with empty = 0? It seems like that might make the operations somewhat more natural (I haven't looked at the implementation, though).

Reply:
I considered doing that too but thought this was more natural. Implementation-wise I think they are roughly equivalent, i.e. `-1` when checking if the pool is empty/full vs `-1` when adding/removing entries. I don't have a strong opinion on this but as I said I find `top` more natural. Please let me know if you would like me to change this to `len`.

Reply:
I don't have a strong feeling either way. Thanks for checking it out.